pkg/pma/pma.c

```diff
@@ -0,0 +1,813 @@
+        err = EINVAL;
+        goto fail;
+    }
+
+    size_t heap_len  = 0;
+    size_t stack_len = 0;
+    if (pma->len_getter(&heap_len, &stack_len) == -1) {
+        err = ECANCELED;
+        fprintf(stderr,
+                "pma: failed to determine length of heap and stack\r\n");
+        goto fail;
+    }
+
+    heap_len  = round_up(heap_len, kPageSz);
+    stack_len = round_up(stack_len, kPageSz);
+
+    wal_t      *wal     = pma->heap_fd != -1 || pma->stack_fd != -1
+                              ? malloc(sizeof(*wal))
+                              : NULL;
+    const char *wal_dir = NULL;
+    if (wal) {
+        // We arbitrarily use the heap file's base directory to house the WAL;
+        // the stack's directory could be used instead. dirname() may modify its
+        // argument, so we give it a copy.
+        char *heap_file_copy = strdup(pma->heap_file);
+        wal_dir              = dirname(heap_file_copy);
+        if (wal_open(wal_dir, wal) == -1) {
+            err = errno;
+            free(heap_file_copy);
+            fprintf(stderr,
+                    "pma: failed to open wal file at %s: %s\r\n",
+                    wal_dir,
+                    strerror(err));
+            goto fail;
+        }
+        free(heap_file_copy);
+    }
+
+    if (pma->heap_fd != -1) {
+        assert(wal);
+        if (_append_dirty_pages(wal,
+                                pma->heap_start,
+                                false,
+                                pma,
+                                heap_len)
+            == -1)
+        {
+            err = ECANCELED;
+            fprintf(stderr,
+                    "pma: failed to sync heap changes to %s\r\n",
+                    pma->heap_file);
+            goto fail;
+        }
+    }
+
+    if (pma->stack_fd != -1) {
+        assert(wal);
+        if (_append_dirty_pages(wal,
+                                pma->stack_start,
+                                true,
+                                pma,
+                                stack_len)
+            == -1)
+        {
+            err = errno;
+            fprintf(stderr,
+                    "pma: failed to sync stack changes to %s: %s\r\n",
+                    pma->stack_file,
+                    strerror(err));
+            goto fail;
+        }
+    }
+
+    // Sync and apply the WAL if the backing files are non-NULL.
+    if (wal) {
+        if (wal_sync(wal) == -1) {
+            err = errno;
+            fprintf(stderr,
+                    "pma: failed to sync wal for %s: %s\r\n",
+                    wal_dir,
+                    strerror(err));
+            goto fail;
+        }
+
+        if (wal_apply(wal, pma->heap_fd, pma->stack_fd) == -1) {
+            err = errno;
+            fprintf(stderr,
+                    "pma: failed to apply wal to %s: %s\r\n",
+                    wal_dir,
+                    strerror(err));
+            goto fail;
+        }
+
+        wal_destroy(wal);
+    }
```

Comment:
@ashelkovnykov asked what happens to clean file-backed pages that fall within the truncated region of the file. In other words, what happens when the part of a file backing a mapping is deleted? I wasn't sure, so I looked it up in the `mmap()` man page, which says,

> If the size of the mapped file changes after the call to mmap() as a result of some other operation on the mapped file, the effect of references to portions of the mapped region that correspond to added or removed portions of the file is unspecified.

By not truncating the file, we increase the number of pages in the loom that are file-backed, thereby reducing the number of anonymous pages the kernel needs to allocate.

Reply:
I just realized during my weekly sync w/ @joemfb that we need to explicitly change the permissions of the leftover pages when the size of the heap / stack shrink. Otherwise, this chain of events is possible:
1. `mmap` in initial heap
    1. For example purposes, let's say it's 1 GiB in size
    2. All pages mapped `PROT_READ`, `MAP_FIXED | MAP_PRIVATE`
3. Attempt to write to address at `rut` + 500 MiB
4. `SEGFAULT`; initial page status in `pma->pg_status` bitmap is `PS_MAPPED_CLEAN`
5. Handle fault by calling `mprotect` with new permissions `PROT_READ | PROT_WRITE` for page; page status in `pma->pg_status` bitmap set to `PS_MAPPED_DIRTY`
6. Attempt to write to address at `rut` + 500 MiB again; success
8. `pma_sync`
    1. Let's say that the heap shrank to 500 MiB, as reported by `pma->len_getter`
    2. Changes to page at `rut` + 500 MiB get ignored by the WAL (since it's outside the new heap range), but the page remains mapped with `PROT_READ | PROT_WRITE` permissions
    3. Heap is remapped from backing file; only the mappings **up to** `rut` + 500 MiB are updated 
    4. Page status in `pma->pg_status` bitmap is reset to `PS_UNMAPPED`
9. `hat` advances; attempt to write to address at `rut` + 500 MiB
10. No `SEGFAULT`; write succeeds because permissions are still `PROT_READ | PROT_WRITE`
    1. Thus, `pma->pg_status` never sets bitmap entry for page at `rut` + 500 MiB to `PS_MAPPED_DIRTY`
11. `pma_sync`
    1. Let's say that the heap grows to 750 MiB, as reported by `pma->len_getter`
    2. Attempt to add changes to WAL
    3. Crash in `_append_dirty_pages` because `pma->pg_status` bitmap still set to `PS_UNMAPPED`

The above chain of events is entirely possible from the perspective of the PMA. This may not be an issue for Vere if there's a guarantee that Vere will never:
a. write to heap / stack; AND
b. shrink heap / stack to less than the size of a written address

However, the PMA is being treated as an independent lib (hence its own style and conventions) and not a Vere component, so the error needs to be resolved.

**TL;DR:**
Need to `mprotect` addresses in range from new `hat` to old `hat` as `PROT_NONE` when shrinking heap (same goes for the stack, but with range old `cap` to new `cap`)

Reply:
If the dirty page falls within the bounds of a backing file but outside the range of the new heap length (or stack length), then this problem doesn't arise because we re-map the backing file as read-only at the end of `pma_sync()`, which clobbers the existing mappings. This is a problem though for dirty pages that fall outside the bounds of a backing file. To rectify this, I added logic at the end of `pma_sync()` which marks dirty pages in the space not backed by a file as read-only. Note this wasn't an issue originally because we previously unmapped the free space at the end of `pma_sync()` until that was removed in 3f4a31eb41026418a99f29328ba8121139ca4324.